ruby2d/ruby2d
# Canvas refactor using sdl renderer

@blacktm, this refactors Canvas implementation to use `SDL_Renderer` since it supports a software-based renderer with alpha blending support, and provides some drawing primitives that we don't have to reinvent. 

* Added following fill methods
  * `fill_triangle` and `fill_quad` methods for a single triangle and quadrilateral respectively, with vertex colouring. 
  * `fill_circle` with a single colour
  * `fill_ellipse` with a single colour
* Added folowing draw methods
  * `draw_circle` with "pen width" support
    * Renamed `width` -> `pen_width` to avoid ambiguity with shape dimension in other methods
  * `draw_polyline` with "pen_width" support 
    * with `closed:` param that can be set to `true` to draw polygon outlines
  * `draw_triangle` and `draw_quad` with pen width (using closed polylines)
  * `draw_ellipse` with pen width
* Refactored methods
  * `draw_line` to handle thick lines
    * Renamed `width` -> `pen_width` to avoid ambiguity with shape dimension in other methods
  * `draw_rectangle` to support pen width
  * Replaced `opts={}` params with explicit keyword params
* Updated `canvas` test: 
  * use "space" to enable/disable clearing between frames
  * use "s" key to change the shape following the mouse between square and circle  
* Added tests
  * `canvas_blend` test to verify all the supported shape methods
  * `canvas_linejoin` interactive test to verify `polyline` mitre joining by controlling each point with mouse (use number keys to select point)

> FYI this has been tested with `ruby` and `mruby` on my M1-based Macbook and not on any other platforms. 

<img width="657" alt="Screen Shot 2022-04-28 at 18 12 36" src="https://user-images.githubusercontent.com/19142777/165856410-3b5c91d9-ec31-4dbf-987e-2b0cd6a329ec.png">
